### PR TITLE
Change Ray's default port to avoid conflicts with Redis

### PR DESCRIPTION
**Pull request by infwinston**
Fixes https://github.com/skypilot-org/skypilot/issues/1789. I looked up common ports used by well-known services here
https://en.wikipedia.org/wiki/List_of_TCP_and_UDP_port_numbers
and 6380 seems safe.

This PR also enables Ray applications on SkyPilot by making our Ray cluster invisible by changing the default `temp_dir` to `/tmp/ray_skypilot`. (related https://github.com/skypilot-org/skypilot/issues/1305, https://github.com/skypilot-org/skypilot/issues/671)

Tested (run the relevant ones):

- [x] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

### CODE REVIEW

**Comment by suquark** [comment]
we should use `ray_skypilot` instead of `ray_patch` as the temp dir name, in case this becomes our long-term solution

File: sky/templates/azure-ray.yml.j2
Comment (concretevitamin) [inline sky/templates/azure-ray.yml.j2:146]: Is it ok that our Ray continues to use `--object-manager-port=8076`?

**Review by concretevitamin — COMMENTED** [review]
Awesome! Quick look. `ray start -h` shows some other ports to check. Is it ok to share them with an app's Ray?
```
  --object-manager-port INTEGER   the port to use for starting the object
                                  manager
  --node-manager-port INTEGER     the port to use for starting the node
                                  manager
  --gcs-server-port INTEGER       Port number for the GCS server.
  --min-worker-port INTEGER       the lowest port number that workers will
                                  bind on. If not set, random ports will be
                                  chosen.
  --max-worker-port INTEGER       the highest port number that workers will
                                  bind on. If set, '--min-worker-port' must
                                  also be set.
  --worker-port-list TEXT         a comma-separated list of open ports for
                                  workers to bind on. Overrides '--min-worker-
                                  port' and '--max-worker-port'.
  --ray-client-server-port INTEGER
                                  the port number the ray client server binds
                                  on, default to 10001.
...
  --dashboard-port INTEGER        the port to bind the dashboard server to--
                                  defaults to 8265
  --dashboard-agent-listen-port INTEGER
                                  the port for dashboard agents to listen for
                                  http on.
  --dashboard-agent-grpc-port INTEGER
                                  the port for dashboard agents to listen for
                                  grpc on.
  --metrics-export-port INTEGER   the port to use to expose Ray metrics
                                  through a Prometheus endpoint.
```
May want to run
- https://github.com/skypilot-org/skypilot/tree/master/examples/ray_tune_examples
- AutoGluon
- back compat

**Comment by infwinston** [comment]
> Is it ok that our Ray continues to use --object-manager-port=8076?

To my understanding, from their [doc](https://docs.ray.io/en/latest/ray-core/configure.html#all-nodes),`object-manager-port` by default will be randomly decided in runtime (in which user's ray will use a port different from `8076`).
so it'd not cause issue. But @Michaelvll do you know why we use a specific port `8076` in the first place?

I'll upload an AutoGluon example I run before.

Back compat should be fine if we keep our dashboard running on the same port `8265`, which is used to submit our Ray job. However, if user's Ray app requires dashboard, it'll still be conflicts. We may need to decide whether to change our dashboard port. each side comes with pros/cons.

Btw, I'm also blocked by this issue in vicuna project in which I use ray to distribute work to multiple gpus. Now my sky yaml fails to run my app. We may want to merge this PR soon.

**Comment by iojw** [comment]
Tested and confirmed that the following backwards compatibility cases work:
- `exec` on existing up cluster - as discussed, this depends on the job manager port, so changing the GCS port does not affect the submission of jobs
- `start` on existing stopped cluster - this re-provisions the cluster using the new GCS port, so there is no issue here

**Comment by Michaelvll** [comment]
A kind bump for this PR @infwinston.

**Comment by infwinston** [comment]
Will get back to office and look into this tomorrow.

**Comment by infwinston** [comment]
I've updated this PR to ensure backward compatibility! and tested it.

**Review by Michaelvll — COMMENTED** [review]
Thanks for the fix @infwinston! This would be very useful for users with their own ray cluster. Left several comments.

File: sky/backends/backend_utils.py
Comment (infwinston) [reply sky/backends/backend_utils.py:856]: Ah I tried this before. The issue is `json.dumps` will add a space between ":" and the value. which will cause the YAML parser to fail... we need `"{'ray_port':6380}"` not `"{'ray_port': 6380}"`..

I've added some comments to explain it.

File: sky/backends/cloud_vm_ray_backend.py
Comment (infwinston) [reply sky/backends/cloud_vm_ray_backend.py]: Good point! updated.

File: sky/skylet/constants.py
Comment (infwinston) [reply sky/skylet/constants.py:10]: Added!

File: sky/backends/cloud_vm_ray_backend.py
Comment (infwinston) [reply sky/backends/cloud_vm_ray_backend.py]: Good point. Added!

**Comment by infwinston** [comment]
I've merged the master and addressed the comment.

Tested:
- `sky launch examples/autogluon.yaml` works.
- Backward compatibility: I switched to master and launched a VM. and switch back to this branch. `sky exec` still works.

@Michaelvll could you review if there's anything left? thanks!

File: sky/backends/backend_utils.py
Comment (Michaelvll) [reply sky/backends/backend_utils.py:856]: Ahh, got it. If we add a `|` after the `-` in the `head_start_ray_commands`, the problem should be fixed. : )

File: sky/backends/cloud_vm_ray_backend.py
Comment (Michaelvll) [inline sky/backends/cloud_vm_ray_backend.py:237]:
```suggestion
            # Only set the `_temp_dir` to SkyPilot's ray cluster directory when the directory
            # exists for backward compatibility for the VM launched before #1790.
            if os.path.exists({constants.SKY_REMOTE_RAY_TEMPDIR!r}):
```

**Review by Michaelvll — APPROVED** [review]
Thanks for the fix @infwinston! LGTM. It should be good to go if the tests are passed.

**Comment by infwinston** [comment]
Thanks a lot for the reviews @Michaelvll! Now all the smoke tests passed. Merging this.

**Pull request merged**
